005-1-locations-on-stack.md

```diff
@@ -0,0 +1,590 @@
+> 
+> - Memory locations, for values that are stored in memory, starting
+>   at a given address within a given address space.
+> 
+> - Register locations, for values that are stored in registers.
+> 
+> - Undefined locations, for values that are unavailable
+>   (perhaps due to optimization).
+> 
+> - Implicit locations, for values that are not stored anywhere (like
+>   undefined locations), but whose values can be reconstructed.
+> 
+> - Composite locations, for values that are stored in a combination
+>   of storage (e.g., aggregate values where certain components have
+>   been promoted to registers).
+> 
+> A DWARF expression is encoded as a stream of operations, each consisting of an
+> opcode followed by zero or more literal operands. The number of operands is
+> implied by the opcode.
+> 
+> The result of a DWARF expression is the value or location on the top
+> of the stack after evaluating the operations.
+> 
+> A DWARF expression is evaluated in a context that determines whether
+> its result is expected to be a value or a location. Expressions that are
+> expected to produce a location are called "location descriptions."
+> 
+> Implicit conversions between memory addresses and values may happen
+> during the execution of any operation or when evaluation of the
+> expression is completed. If a location is expected, but the result is
+> a value, the value is implicitly treated as a memory address in the
+> default address space, and converted to a memory location. If a value
+> is expected, but the result is a memory location in the default
+> address space, the address is implicitly converted to a value.
+
+
+### Section 2.5.1 General Operations
+
+[Page 26]
+In the first paragraph, delete all but the first sentence of the first
+two paragraphs, leaving only this:
+
+> Each general operation represents a postfix operation on a simple
+> stack machine.
+
+(These deleted sentences have been moved up to Section 2.5.)
+
+
+### Section 2.5.1.1 Literal Encodings
+
+Move `DW_OP_addr` and `DW_OP_addrx` to Section 2.5.2.2.
+
+
+### Section 2.5.1.3 Stack Operations
+
+[Pages 30-31]
+Under `DW_OP_deref`, `DW_OP_deref_size`, and `DW_OP_deref_type`,
+replace the descriptions with:
+
+> 7\. `DW_OP_deref`  
+> The `DW_OP_deref` operation pops the top stack entry and treats it as
+> a location. The first `S` bytes, where `S` is the size of an address on
+> the target machine, are retrieved from the location and pushed onto
+> the stack as a value of the generic type.
+> 
+> 8\. `DW_OP_deref_size`  
+> The `DW_OP_deref_size` takes a single 1-byte unsigned integral operand
+> that specifies the size `S`, in bytes, of the value to be retrieved. The
+> operation behaves like the `DW_OP_deref` operation: it
+> pops the top stack entry and treats it as a location. The first `S` bytes
+> are retrieved from the location, zero extended to the size of an
+> address on the target machine, and pushed onto the stack as a value of
+> the generic type.
```

Comment:
What if `S` is bigger than the size of an address? DWARF-5 says "whose value may not be larger than the size of the generic type". This information seems to have been lost.